sim/midas/src/main/cc/bridges/cpu_managed_stream.cc

```diff
@@ -0,0 +1,80 @@
+#include "cpu_managed_stream.h"
+
+#include <iostream>
+#include <assert.h>
+
+/**
+ * @brief Enqueues as much as num_bytes of data into the associated stream
+ *
+ * @param src Source from which to copy data to enqueue
+ * @param num_bytes Desired number of bytes to enqueue
+ * @param threshold_bytes Minimum number of bytes to enqueue. If fewer bytes
+ *        would be enqueued, this method enqueues none and returns 0.
+ * @return size_t
+ */
+size_t StreamFromCPU::push(void* src, size_t num_bytes, size_t threshold_bytes) {
+  assert(num_bytes >= threshold_bytes);
+
+  // Similarly to above, the legacy implementation of DMA does not correctly
+  // implement non-multiples of 512b. The FPGA-side queue will take on the
+  // high-order bytes of the final beat in the transaction, and the strobe is
+  // not respected. So put the assertion here and discuss what to do next.
+  assert((num_bytes % DMA_BEAT_BYTES) == 0);
+
+  auto num_beats = num_bytes / DMA_BEAT_BYTES;
+  auto threshold_beats = threshold_bytes / DMA_BEAT_BYTES;
+
+  assert(threshold_beats <= this->fpga_buffer_size());
+  auto space_available = this->fpga_buffer_size() - this->mmio_read(this->count_addr());
+
+  if ((space_available == 0) || (space_available < threshold_beats)) {
+    return 0;
+  }
+
+  auto push_beats = std::min(space_available, num_beats);
+  auto push_bytes = push_beats * DMA_BEAT_BYTES;
+  auto bytes_written = pcis_write(this->dma_addr(), (char*) src, push_bytes);
+  assert(bytes_written == push_bytes);
+
+  return bytes_written;
+}
+
+/**
+ * @brief Dequeues as much as num_bytes of data from the associated bridge stream.
+ *
+ * @param dest  Buffer into which to copy dequeued stream data
+ * @param num_bytes  Bytes of data to dequeue
+ * @param threshold_bytes  Minimum number of bytes to dequeue. If fewer bytes would
+ *        be dequeued, dequeue none and return 0.
```

Comment:
There is currently no case where it is not `0` or `num_bytes`. Given that, it might be natural to simply provide a boolean flag (`batch`) that ensures the function returns `num_bytes` or `0`. We could also just have two different pairs of methods (`pull` and `batch_pull`) or something like that. 

That aside, `required_bytes` seems like a better name to me. I'll change it.

Reply:
I think that just changing the param name to `required_bytes` is fine.  I also thought about splitting the methods or making it a bool and I think renaming it addresses my primary concern.